source/CantorSchroederBernstein/CSB-WLPO.lagda

```diff
@@ -0,0 +1,286 @@
+open import UF.Embeddings
+open import UF.Equiv
+open import NotionsOfDecidability.Complemented
+open import NotionsOfDecidability.Decidable
+open import UF.Retracts
+
+\end{code}
+
+In this file we consider a generalization of the Cantor–Schröder–Bernstein (CSB)
+theorem only assuming the weak limited principle of omniscience (WLPO), based on
+König's argument.
+
+Theorem.
+Assuming WLPO, then for every pair of types X and Y, if there are complemented
+embeddings X ↪ Y and Y ↪ X, then X is equivalent to Y.
+
+In particular, we do not assume function extensionality.
+
+This theorem can be viewed as a proper generalization of the
+Cantor–Schröder–Bernstein theorem to arbitrary non-topological ∞-topoi, since,
+under the assumption of the law of excluded middle (LEM), every embedding is
+complemented. On the other hand, It was shown by Pradic and Brown (2022) that
+the Cantor–Schröder–Bernstein theorem in its most naïve form implies the law of
+excluded middle:
+
+If it is true that for every pair of sets X and Y, if X injects into Y and Y
+injects into X then X and Y are in bijection, then the law of excluded middle
+holds.
+
+Hence we also know that in the absence of the law of excluded middle, the
+hypotheses of the theorem must be strengthened.
+
+Construction
+------------
+
+For our formalization we import a series of properties of perfect images from
+which the construction is straight forward.
+
+\begin{code}
+
+module _ {X : 𝓤 ̇ } {Y : 𝓥 ̇ } {f : X → Y} {g : Y → X} where
+
+ map-construction : (x : X) → is-decidable (is-perfect-image f g x) → Y
+ map-construction x (inl γ) = inverse-on-perfect-image x γ
+ map-construction x (inr nγ) = f x
+
+ map-construction-CSB :
+  ((x : X) → is-decidable (is-perfect-image f g x)) → X → Y
+ map-construction-CSB D x = map-construction x (D x)
+
+ map-construction-is-left-cancellable :
+  (lc-f : left-cancellable f) →
+  {x x' : X}
+  (d : is-decidable (is-perfect-image f g x))
+  (d' : is-decidable (is-perfect-image f g x')) →
+  map-construction x d ＝ map-construction x' d' → x ＝ x'
+ map-construction-is-left-cancellable lc-f {x} {x'} (inl ρ) (inl ρ') p =
+  inverse-on-perfect-image-is-section x ρ ⁻¹
+  ∙ ap g p
+  ∙ inverse-on-perfect-image-is-section x' ρ'
+ map-construction-is-left-cancellable lc-f {x} {x'} (inl ρ) (inr nρ') p =
+  𝟘-elim (perfect-image-is-disjoint x' x nρ' ρ (p ⁻¹))
+ map-construction-is-left-cancellable lc-f {x} {x'} (inr nρ) (inl ρ') p =
+  𝟘-elim (perfect-image-is-disjoint x x' nρ ρ' p)
+ map-construction-is-left-cancellable lc-f {x} {x'} (inr nρ) (inr nρ') = lc-f
+
+ map-construction-CSB-is-left-cancellable :
+  left-cancellable f →
+  (D : (x : X) → is-decidable (is-perfect-image f g x)) →
+  left-cancellable (map-construction-CSB D)
+ map-construction-CSB-is-left-cancellable lc-f D {x} {x'} =
+  map-construction-is-left-cancellable lc-f (D x) (D x')
+
+\end{code}
+
+Computations with the construction.
+
+\begin{code}
+
+module _
+ {X : 𝓤 ̇ } {Y : 𝓥 ̇ } {f : X → Y} {g : Y → X}
```

Comment:
See how we indent parameters in anonymous modules by looking at examples in TypeTopology.